### ISSUE TITLE
[ASTextNode] Calling calculateSizeThatFits: has side effects

### ISSUE THREAD

**Issue by maicki** [facebook/AsyncDisplayKit#2593]
Calling `calculateSizeThatFits` within `ASTextNode` has a side effect of updating the `_constrainedSize` variable of itself with the passed in constrained size.

This is not great as multiple layout passes could happen with a different `constrainedSize` and the final size of the text node be 

One example would be to implement `sizeThatFits:` on a `sizing node` with one `ASTextNode` as subnode. Let's use this code for example:

```Objc
- (void)updateNodeLayout
{
  // Use the bounds of the view and get the fitting size
  CGSize size = [self.sizingNode layoutThatFits:ASSizeRangeMake(CGSizeZero, CGSizeMake(CGFLOAT_MAX, 100.0))].size;
  size.width -= 10;
  self.sizingNode.frame = CGRectMake((self.view.bounds.size.width - size.width) / 2.0,
                                           (self.view.bounds.size.height - size.height) / 2.0,
                                            size.width,
                                            size.height);
}
```

* The first time calling `sizeThatFits:` one the root node would return a CGSize of:  `(CGSize) size = (width = 58.5, height = 34)`
* Reducing the width of -10: `(CGSize) size = (width = 48.5, height = 34)`
* Applying this size to the sizing node will re-trigger a measurement pass within `__layout` with the new bounds size, what is expected. This will actually also layout the `ASTextNode` with the right truncation as the `sizingNode` need to do a complete measurement pass.
* Some rotation is now happening and the `updateNodeLayout` method will be called again
* The call to `sizeThatFits:` go through all subnodes of the `sizingNode` to calculate the fitting size again. Unfortunately in this process it will call `calculateSizeThatFits:` on the text node with a different constrained size as it was laid out.
* The returned size will be `(CGSize) size = (width = 58.5, height = 34)` and we are reducing the width again with 10 to final size: `(CGSize) size = (width = 48.5, height = 34)`
* This time if we set the frame again to the `sizingNode` no measurement pass is happening again as no layout was invalidated or the bounds changed. This has the effect that the `_constrainedSize` of the `ASTextNode` has the wrong value

One workaround in the example above is to set `[self.sizingNode setNeedsLayout]` before setting the `frame`, but this is only the right way as the `ASTextNode` is an immediate subnode. If it would be another hierarchy deeper it would not work and the `ASTextNode`'s `_constrainedSize` will not have the correct value.

**Comment by maicki** [comment]
I will assign this issue to @Adlai-Holler as I think he's working on `ASTextNode` things

**Comment by Adlai-Holler** [comment]
@maicki Is this a blocker for the layout coalescing diff? If not, we should remove it from the milestone and focus on other stuff for 2.0 since it's not an API change.

**Comment by maicki** [comment]
@Adlai-Holler I think it's not a blocker for #2597 per-se as `ASTextNode` in it's current state has this side effects. So there should be no regression if you use ASDK containers the way as before.

That said with the diff mentioned above we now support also manual layout of nodes, in that case if a text node is involved it can happen, that in the measurement pass the constrained size instance variable will get a different constrained size as the final one as a different frame will be set. In that case `ASTextNode` will not work the way it should ...

**Comment by maicki** [comment]
This should be resolved with #2643 and small follow up #2726
